tests/helpers.py

```diff
@@ -160,3 +166,174 @@ def __init__(self, **kwargs):
+            validation
+        - An exception to expect during deserialization and the bytes to deserialize
+
+    Example usage:
+    ```python
+    @final
+    @dataclass
+    class ToyClass(Serializable):
+        a: int
+        b: str
+
+        def serialize_to(self, buf: Buffer):
+            buf.write_varint(self.a)
+            buf.write_utf(self.b)
+
+        @classmethod
+        def deserialize(cls, buf: Buffer) -> "ToyClass":
+            a = buf.read_varint()
+            b = buf.read_utf()
+            if len(b) > 10:
+                raise ValueError("b must be less than 10 characters")
+            return cls(a, b)
+
+        def validate(self) -> None:
+            if self.a == 0:
+                raise ZeroDivisionError("a must be non-zero")
+
+    gen_serializable_test(
+        context=globals(),
+        cls=ToyClass,
+        fields=[("a", int), ("b", str)],
+        test_data=[
+            ((1, "hello"), b"\x01\x05hello"),
+            ((2, "world"), b"\x02\x05world"),
+            ((0, "hello"), ZeroDivisionError),
+            (IOError, b"\x01"), # Not enough data to deserialize
+            (ValueError, b"\x01\x0bhello world"), # 0b = 11 is too long
+        ],
+    )
+    ```
+    This will add 1 class test with 4 test functions containing the tests for serialization, deserialization,
+    validation, and deserialization error handling
+
+
+    """
+    # Separate the test data into parameters for each test function
+    # This holds the parameters for the serialization and deserialization tests
+    parameters: list[tuple[dict[str, Any], bytes]] = []
+
+    # This holds the parameters for the validation tests
+    validation_fail: list[tuple[dict[str, Any], type[Exception]]] = []
+
+    # This holds the parameters for the deserialization error tests
+    deserialization_fail: list[tuple[bytes, type[Exception]]] = []
+
+    for data_or_exc, expected_bytes_or_exc in test_data:
+        if isinstance(data_or_exc, tuple) and isinstance(expected_bytes_or_exc, bytes):
+            kwargs = dict(zip([f[0] for f in fields], data_or_exc))
+            parameters.append((kwargs, expected_bytes_or_exc))
+        elif isinstance(data_or_exc, type) and isinstance(expected_bytes_or_exc, bytes):
+            deserialization_fail.append((expected_bytes_or_exc, data_or_exc))
+        else:
+            data = cast(Tuple[Any, ...], data_or_exc)
+            exception = cast(type[Exception], expected_bytes_or_exc)
+            kwargs = dict(zip([f[0] for f in fields], data))
+            validation_fail.append((kwargs, exception))
+
+    def generate_name(param: dict[str, Any] | bytes, i: int) -> str:
+        """Generate a name for the test case."""
+        length = 30
+        result = f"{i:02d}] : "  # the first [ is added by pytest
+        if isinstance(param, bytes):
+            try:
+                result += str(param[:length], "utf-8") + "..." if len(param) > (length + 3) else param.decode("utf-8")
```

Comment:
hmm, I'm not so sure about the `param.decode("utf-8")` here. Why are we decoding the bytes into utf-8? Why not just always use `repr`? Sometimes this might make sense yeah, but other times, it can happen to succeed even when it really doesn't make sense (the data isn't meant to be utf-8 text, yet we decode it as such). I think let's just do the `repr`.